File: version4j/src/main/java/com/only4/algorithm/version4j/leetcode/SingleNumber.java
package com.only4.algorithm.version4j.leetcode;

public class SingleNumber {
    public int singleNumber(int[] nums) {
        // 利用异或运算性质：a ^ a = 0, a ^ 0 = a
        // 相同数字异或会抵消，只剩下出现一次的数字
        int uniqueNumber = nums[0];

        for (int arrayIndex = 1; arrayIndex < nums.length; arrayIndex++) {
            int currentNumber = nums[arrayIndex];
            uniqueNumber = uniqueNumber ^ currentNumber;
        }

        return uniqueNumber;
    }
}

File: version4k/src/main/kotlin/com/only4/algorithm/version4k/leetcode/singleNumber.kt
package com.only4.algorithm.version4k.leetcode

fun singleNumber(nums: IntArray): Int {
    // 利用异或运算性质：a xor a = 0, a xor 0 = a
    // 相同数字异或会抵消，只剩下出现一次的数字
    var uniqueNumber = nums[0]

    for (arrayIndex in 1 until nums.size) {
        val currentNumber = nums[arrayIndex]
        uniqueNumber = uniqueNumber xor currentNumber
    }

    return uniqueNumber
}
